---
title:"Figure 1B tsne"
author: "SJ"
output: html_document
---

```
set.seed(199)
rm(list = ls())

#applied filtering criteria to exclude probable germline genetic variants:
#allelic depth greater than 50 and allelic fraction greater than 0.25
#If variants were not in Cosmic database version 70, we excluded those with allele frequency in 1000 Genomes Phase 3 version 5 (August 2015) or ExAC database greater than 0.1%.
#retained variants with allele frequency less than 0.1% in ESP 6500si version 2.
#included only exonic or splicing variants with following annotations in the analysis: frameshift deletion, frameshift insertion, missense SNV, nonframeshift deletion, nonframeshift insertion, stopgain, and stoploss

library(data.table)
# files <- list.files("/udd/repsa/projects/sisp/vcfs/Mutation/SNP/Annotation",pattern = glob2rx("*.GATK.snp.annovar.hg19_multianno.xls.gz"),full.names = T)
# files2 <- list.files("/udd/repsa/projects/sisp/vcfs/Mutation/INDEL/Annotation",pattern = glob2rx("*.GATK.indel.annovar.hg19_multianno.xls.gz"),full.names = T)
# dt0 <- sapply(c(files,files2),function(file){
#   cell <- strsplit(basename(file),".",fixed = T)[[1]][1]
#   dt2 <- fread(paste0('zcat ',file),data.table=F,na.strings = ".")
#   dt2$cell <- cell
#   return(dt2)
#   },simplify = F)
# df <- plyr::rbind.fill(dt0)
# save(df,file="/udd/repsa/projects/sisp/df.RData")
load("/udd/repsa/projects/sisp/df.RData")
#sep out
d <- grepl(",",df$GeneName,fixed = T)
table(d)
df0  <- df[d,]
df1  <- df[!d,]

df0a <- sapply(1:dim(df0)[1],function(x){
  di <- df0[x,]
  g <- strsplit(di$GeneName,",",fixed=T)[[1]]
  f <- strsplit(di$Func,";",fixed=T)[[1]]
  if(length(g)!=length(f))
    f <- rep(f[1],length(g))
  k <- sapply(1:length(g),function(i) {
    di$GeneName <- g[i]
    di$Func <- f[i]
    di
  },simplify=F)
  k <- plyr::rbind.fill(k)
  k
},simplify = F)
df0b <- plyr::rbind.fill(df0a)

df2 <- rbind(df0b,df1)
save(df0,df0b,df1,df2,file="/udd/repsa/projects/sisp/df2.RData")
load("/udd/repsa/projects/sisp/df2.RData")
save(df2,file="/udd/repsa/projects/sisp/df2.only.RData")
levels(factor(df1$Func))
View(df1[df1$Func=="exonic;splicing",])
d <- grepl(";",df1$Func,fixed = T)
View(df1[d,])
e <- df1[d,]
e2 <- df1[!d,]


df0a1 <- sapply(1:dim(e)[1],function(x){
  #x <- 1
  di <- e[x,]
  g <- strsplit(di$GeneName,",",fixed=T)[[1]]
  f <- strsplit(di$Func,";",fixed=T)[[1]]
  if(length(g)!=length(f))
    g <- rep(g[1],length(f))
  k <- sapply(1:length(f),function(i) {
    di$GeneName <- g[i]
    di$Func <- f[i]
    di
  },simplify=F)
  k <- plyr::rbind.fill(k)
  k
},simplify = F)
df0b1 <- plyr::rbind.fill(df0a1)

f <- plyr::rbind.fill(list(df0b,df0b1,e2))
save(f,file="/udd/repsa/projects/sisp/f.RData")

load("/udd/repsa/projects/sisp/f.RData")
levels(factor(f$Func)) # c("exonic","splicing")
levels(factor(f$ExonicFunc)) # c("frameshift deletion","frameshift insertion","missense SNV","nonframeshift deletion","nonframeshift insertion","stopgain","stoploss")
f2 <- f[which((f$Func %in% c("exonic","splicing"))& (f$ExonicFunc %in% c("frameshift deletion","frameshift insertion","missense SNV","nonframeshift deletion","nonframeshift insertion","stopgain","stoploss"))),]

threshold <- 0.001
dt2 <- f2
# #remove common variants in dbSNP (> threshold or NA) but retain if present in COSMIC
dt3 <- dt2[which(is.na(dt2$`1000g2015aug_all`)|(!is.na(dt2$`1000g2015aug_all`)&dt2$`1000g2015aug_all`< threshold)|!is.na(dt2$cosmic70)),]
# #remove common variants in ExAC (> threshold or NA) but retain if present in COSMIC
dt4 <- dt3[which(is.na(dt3$ExAC_ALL)|(!is.na(dt3$ExAC_ALL)&dt3$ExAC_ALL< threshold)|!is.na(dt3$cosmic70)),]
# #only NA or ESP < 0.1%
dt5 <- dt4[which(dt4$esp6500siv2_all<0.001|is.na(dt4$esp6500siv2_all)),]
# 

dt5$DP <- sapply(dt5$INFO,function(x){
#x <- [1]
y <- strsplit(x,"DP=")[[1]][2]
as.numeric(strsplit(y,";",fixed = T)[[1]][1])
})

dt5$VAF <- apply(dt5,1,function(x){
  cell <- x["cell"]
  i <- x[cell]
  y <- strsplit(trimws(i),":")[[1]][2]
  k <- as.numeric(strsplit(y,",",fixed = T)[[1]])
  k[2]/(k[1]+k[2])
})

#allelic fraction and depth
dt6 <- dt5[which(dt5$VAF>0.25&dt5$DP>50),]
#save(dt6,file="/udd/repsa/projects/sisp/dt6.RData")
#load("/udd/repsa/projects/sisp/dt6.RData")
pt <- dt6
library(data.table)
setnames(pt,"GeneName","gene")
setnames(pt,"cell","sample")
setnames(pt,"ExonicFunc","variant_class")
levels(factor(pt$variant_class))
vorder <- c("stopgain","frameshift deletion","frameshift insertion",
            "nonframeshift deletion","nonframeshift insertion","missense SNV",
            "stoploss")
library("scales")
library(ggsci)
#write table 

#write.csv(dt6,file=paste0(glob_wd,"treshold",".csv"),row.names = F)
#pdf(file=paste0(glob_wd,"treshold",".pdf"),width = 12,height = 6)
#waterfall(pt, fileType = "Custom", variant_class_order = vorder, 
#          mainXlabel = TRUE,maxGenes = 40,
#          mainPalette = pal_npg("nrc")(length(vorder)))
#dev.off()

#plot mutationlandscape using complexheatmap 
#plot only significant mutated and found in CCA tissue


library(ComplexHeatmap)

glob_wd = "G:/My Drive/Github/Figure 1/1 mutation landscape"

setwd(glob_wd)
fl.output = paste(glob_wd,"/output",sep = "")
fl.input =  paste(glob_wd,"/input",sep = "")


mat1 <- read.csv("/Onco_only reported Gene.csv", comment.char="#")
mat<- (mat1[,2:16])
rnames <- mat1[,1]  # assign labels in column 1 to "rnames"
rownames(mat)<- rnames



mat$X     <- NULL #del col1

#mat<-t(mat[,1:39])

df= data.frame(Ethnicity = c(rep("Thai", 3), rep("Japanese", 1),rep("Thai", 4),rep("Japanese",2), rep("Thai", 1),rep("Japanese",1), rep("Thai", 1),rep("Japanese",2)))
#df= data.frame(Ethnicity = c(rep("Thai", 9), rep("Japanese", 6)))
ha = HeatmapAnnotation(df = df, col = list(Ethnicity  = c("Thai" =  "brown", "Japanese" = "yellow")),
                       annotation_height = unit(c(3, 3, 5), "mm"),annotation_name_offset = unit(50, "mm"),
                       annotation_legend_param = list(Ethnicity = list(title = "Ethnicity",
                                                                       title_position = "leftcenter",
                                                                       nrow = 1 )))
#sort1<-[order(mat1$Order),]
alter_fun = list(
    
    background = function(x, y, w, h) {
        grid.rect(x, y, w-unit(0.5, "mm"),h*0.4, gp = gpar(fill = "ivory2", col = NA))
    },
    
    frameshift_deletion = function(x, y, w, h) {
        grid.rect(x, y, w-unit(0.5, "mm"), h*0.4, gp = gpar(fill = "paleturquoise", col = NA))
    },
    
    frameshift_insertion = function(x, y, w, h) {
        grid.rect(x, y, w-unit(0.5, "mm"), h*0.4, gp = gpar(fill = "dodgerblue3", col = NA))
    },
    
    missense = function(x, y, w, h) {
        grid.rect(x, y, w-unit(0.5, "mm"), h*0.4, gp = gpar(fill = "red", col = NA))
    },
    
    nonframeshift_deletion = function(x, y, w, h) {
        grid.rect(x, y, w-unit(0.5, "mm"), h*0.4, gp = gpar(fill = "darkcyan", col = NA))
    },
    
    nonframeshift_insertion = function(x, y, w, h) {
        grid.rect(x, y, w-unit(0.5, "mm"), h*0.4, gp = gpar(fill = "mediumorchid1", col = NA))
    },
    
    stopgain = function(x, y, w, h) {
        grid.rect(x, y, w-unit(0.5, "mm"), h*0.4, gp = gpar(fill = "blue", col = NA))
    },
    
    stoploss = function(x, y, w, h) {
        grid.rect(x, y, w-unit(0.5, "mm"), h*0.4, gp = gpar(fill = "Ivory 3", col = NA))
    },
    SNVs_Indel = function(x, y, w, h) {
        grid.rect(x, y, w-unit(0.5, "mm"), h*0.4, gp = gpar(fill = "royalblue1", col = NA))
    }
    
    
    
)



col = c("frameshift_deletion" = "paleturquoise", "frameshift_insertion" = "dodgerblue3", "missense" = "red",
        "nonframeshift_deletion" = "darkcyan", "nonframeshift_insertion" = "darkcyan", "stopgain" = "blue",
        "stoploss" = "Ivory 3","SNVs_Indel" = "royalblue1")

Ht= oncoPrint(mat,
              get_type = function(x) strsplit(x, ";")[[1]],
              alter_fun = alter_fun,
              col = col, row_order = mat1$Order,
              column_title = "Mutation landscape in Cholangiocarcinoma cell lines",
              column_order = NULL,
              show_pct=TRUE,
              show_column_names = TRUE,
              #row_barplot_width = unit(7, "cm"),
              row_names_gp = gpar(fontsize = 9), 
              column_names_gp = gpar(fontsize = 9),
              bottom_annotation = ha, width=unit(10, "cm"),
              remove_empty_columns = FALSE,
              heatmap_legend_param = list(title = "Alternations", at = c("frameshift_deletion", "frameshift_insertion", "missense",
                                                                         "nonframeshift_deletion", "nonframeshift_insertion", "stopgain","stoploss","SNVs_Indel"), 
                                          labels = c("frameshift_deletion", "frameshift_insertion", "missense",
                                                     "nonframeshift_deletion", "nonframeshift_insertion", "stopgain","stoploss","SNVs_Indel"),
                                          nrow = 1,
                                          title_position = "leftcenter"
                                              
                                          
     ))
#heatmap_legend_side = "bottom",   #Alteration
#annotation_legend_side = "bottom" #Ethnic

plotML<-draw(Ht, 
     heatmap_legend_side = "bottom",   
     annotation_legend_side = "bottom" )

decorate_annotation("Ethnicity", {
    grid.text("Ethnicity ", unit(1, "npc") + unit(2, "mm"), 0.5, default.units = "npc", just = "left")
})

pdf(file = paste0(fl.output,"/mutation_landscape",Sys.Date(),".pdf"),width = 13,height = 9)
plot
dev.off()

```


