---
title: "Figure 1A tsne"
author: "PB"
output: html_notebook
---

###set workspace and library###
```{r Introduction, message=FALSE, warning=FALSE, paged.print=FALSE}

######################/2_ccle#####################
##################################################

  graphics.off()
  rm(list=ls())
  glob_wd = "G:/My Drive/Milk CCA/All Final for paper/Figure/High Resolution/final/Draft 28/Code ocean/data"
  setwd(glob_wd)
  set.seed(1234)
  
  collection_name = "Our cell line panel"
  
  path = "/2_tsne"
  wd = "/sisp-ccadr-transcriptomic-analysis"
  my_wd = paste(glob_wd,wd,path,sep = "")
  fl.output = paste(my_wd,"/output",sep = "")
  fl.input =  paste(my_wd,"/input",sep = "")
  set.seed(1234)
  
################################################## 
######################library##################### 
##################################################  
  
  library("openxlsx")
  library("dplyr")
  library("tidyverse")
  library("tidyr")
  library("psych")
  library("caret")
  library("readr")
  library("ggpubr")
  library("ggrepel")
  library("Rtsne")
  library("biomaRt")
  
  #library("org.Hs.eg.db")
  library("sva")
  library("pamr")
  library("ROCR")
  library("caret")
  library("randomForest")
  library("e1071")
  
##################################################  
```

###load metadata###
```{r Meta_data, message=FALSE, warning=FALSE, paged.print=FALSE}
##################################################  
####################meta_data##################### 
##################################################  
  
  metadata_master = as.tibble(openxlsx::read.xlsx(paste0(fl.input
                                                         ,"/_metadata/"
                                                         ,"metadata_master"
                                                         ,".xlsx")))
  metadata_cellline = as.tibble(openxlsx::read.xlsx(paste0(fl.input
                                                         ,"/_metadata/"
                                                         ,"metadata_cellline"
                                                         ,".xlsx")))
  
  metadata_basepair = as.tibble(openxlsx::read.xlsx(paste0(fl.input
                                                           ,"/_metadata/"
                                                           ,"metadata_basepair"
                                                           ,".xlsx")))
  metadata_master = metadata_master %>% filter(validation, ccadr_project)
```

###load ccle data and combination###
```{r load ccle data, message=FALSE, warning=FALSE, paged.print=FALSE}
#metadata
  meta_organ = read_csv(paste0(fl.input
                               ,"/EMBL_EBI_organ.csv"))
  
  meta_organ_t = meta_organ %>% 
    mutate(full_name = paste0(`Factor Value[cell line]`
                              ,", "
                              ,`Factor Value[disease]`)) %>% 
    dplyr::select(full_name,organ,X1) %>% 
    mutate(database = "CCLE"
           ,abbv_name = X1) %>% 
    dplyr::select(-X1) %>% 
    dplyr::select(full_name,organ,database,abbv_name) %>%
    rbind(.
          ,metadata_master %>% 
            dplyr::filter(validation,ccadr_project) %>%
            filter(project %in% "SO_6831") %>%
            filter(treatment %in% c("no","none")) %>% 
            dplyr::select(cell_treatment
                          ,cellline
                          ,abbv_name) %>% 
            distinct() %>%
            mutate(full_name = cell_treatment) %>%
            dplyr::select(full_name
                          ,cellline
                          ,abbv_name) %>%
            left_join(.
                      ,metadata_cellline) %>%
            mutate(database = collection_name) %>%
            dplyr::select(full_name,organ,database,abbv_name) 
          
    )
  
  #sisp collection
  dat_tpm = read_csv(paste0(glob_wd,wd,"/1_cca/output/median_log2tpm_combat_PB_.csv"))

  
  sisp_basal_dat = dat_tpm 
  rm(dat_tpm)
  sisp_basal_dat
  dim(sisp_basal_dat) 
  
  #ccle
  
  dat_ccle <- read_tsv(paste0(fl.input
                              ,"/E-MTAB-2770-query-results.tpms.tsv")
                       ,skip = 4) #934 cell lines
  
  dat_ccle_t = dat_ccle %>% 
                gather(key = celline, value = exp
                       ,-`Gene ID`,-`Gene Name`) %>% 
                mutate(Symbol = `Gene Name`) %>%
                dplyr::select(`Gene ID`,Symbol,celline,exp) %>%
                replace_na(list(exp = 0)) %>% 
                mutate(l_exp = log2(exp+1)) %>% 
                dplyr::select(-exp) %>% 
                group_by(`Gene ID`) %>% 
                mutate(med_l_exp = median(l_exp)) %>%
                filter(med_l_exp > 0) %>% #26,872 gene Symbol
                mutate(ensemble = `Gene ID`) %>%
                group_by() %>%
                dplyr::select(-`Gene ID`,-med_l_exp)  %>% dplyr::select(ensemble
                                                          ,Symbol
                                                          ,celline
                                                          ,l_exp) %>% 
                group_by(Symbol,ensemble) %>%
                spread(key = celline
                       , value = l_exp)

  ##combine ccle to sisp
  #convert sisp to ensemble #hg19
  ensembl = useMart("ENSEMBL_MART_ENSEMBL" #"ensembl"
                    ,dataset="hsapiens_gene_ensembl"
                    ,host = "www.ensembl.org"
                    )
  bm = getBM(attributes= c('ensembl_gene_id'
                           ,'hgnc_symbol'
                           )
        ,filters = 'hgnc_symbol'
        ,values = sisp_basal_dat$Symbol
        ,mart = ensembl
        )
  
  sisp_basal_dat_t = as.tibble(bm) %>% 
                      mutate(Symbol = hgnc_symbol
                             , ensemble = ensembl_gene_id) %>% 
                      dplyr::select(Symbol,ensemble) %>% inner_join(.
                                                                    ,sisp_basal_dat %>% 
                                                                      group_by(Symbol)) %>% #13,543
                      group_by(Symbol,ensemble) %>%
                      gather(key = celline
                             ,value = l_exp
                             ,-Symbol
                             ,-ensemble) %>%
                      spread(key = celline
                             ,value = l_exp) 
  #combat
  combine_dat = inner_join(dat_ccle_t,sisp_basal_dat_t)

  pre_combat = combine_dat %>% 
                group_by() %>% 
                dplyr::select(-ensemble,-Symbol) %>% as.matrix()
  rownames(pre_combat) = combine_dat$ensemble
  
  cl_meta = meta_organ_t %>% 
                      dplyr::select(full_name
                                    ,database) %>% 
                      as.data.frame()
  rownames(cl_meta) = cl_meta$full_name
  cl_meta = cl_meta[colnames(pre_combat)
                    ,]
  
  cl_meta$database = as.factor(cl_meta$database)
  
  modcombat = model.matrix(~database, data=cl_meta)
  edata = as.matrix(pre_combat)
  
  combat_edata = ComBat(dat=edata
                        , batch=cl_meta$database
                        #, mod=modcombat
                        , par.prior=TRUE  
                        , prior.plots=FALSE )
  
  combine_dat_after_combat = as.tibble(cbind(combine_dat %>% 
                                               group_by() %>% 
                                               dplyr::select(ensemble,Symbol)
                                             ,combat_edata))
  
  #normalization
 
  df_combine_dat = combine_dat_after_combat %>% 
                    gather(key = full_name
                           , value = exp
                           , -ensemble, -Symbol) %>% 
                    group_by(full_name) %>% 
                    left_join(.
                              ,meta_organ_t %>% 
                                dplyr::select(full_name, database) ) %>% 
                                group_by(database, ensemble) %>% 
                                mutate(norm_exp = scale(exp)) %>% 
                                group_by() %>% 
                                dplyr::select(ensemble,full_name,norm_exp) %>%
                                spread(key = full_name
                                       , value = norm_exp)
  df_combine_dat = as.data.frame(df_combine_dat)
  rownames(df_combine_dat) = df_combine_dat$ensemble
  df_combine_dat = df_combine_dat[,-1]
```

###run tsne###
```{r}
 dum = cbind(data.frame(full_name = colnames(df_combine_dat)
                         ,stringsAsFactors = FALSE) %>% 
                as.tibble()
              ,t(df_combine_dat) %>% 
                as.tibble()) %>% 
          as.tibble()

  dum2 = (dum %>% 
            dplyr::select(full_name) %>% 
            left_join(meta_organ_t)) %>% 
          left_join(.
                  ,dum)
  
 
  ##check
  dl = as.logical((t(dum2 %>% dplyr::select(-organ,-database,-abbv_name,-full_name) ) == df_combine_dat))
  sum(dl) == length(dl)
  
  #write csv
   write.csv(dum2
            ,paste0(fl.output,"/ccle_cca_expression_table_",Sys.Date(),".csv"))
   
  #annotation for ccle+cca
   result.cca = read_csv(paste0(glob_wd
                                 ,wd
                                 ,"/2_ccle_cca/output/result.cca2019-05-25.csv")) %>% 
                  dplyr::select(-X1)
   meta_organ_t_final =  meta_organ_t %>% 
                            filter(full_name %in% as.character(setdiff(meta_organ_t$full_name
                                                                       ,result.cca$full_name))
                                   ) %>%
                            rbind(.
                                  ,result.cca %>% dplyr::select(colnames(meta_organ_t)))
  #run tsne
  library(Rtsne)
  set.seed(1234)
  dat_tsne = Rtsne::Rtsne(t(df_combine_dat)
                          , dims = 2
                          , perplexity=10
                          , verbose=FALSE
                          , max_iter = 5000
                          , pca_center = TRUE
                          , pca_scale = FALSE
                          
                          ##,theta = 0
  )
  dat_tsne_xy = as.tibble(dat_tsne$Y) %>% 
                  mutate(x = V1, y = V2) %>% 
                  dplyr::select(x,y) %>% 
                  mutate(full_name = colnames(df_combine_dat) )

 
  dum = inner_join(dat_tsne_xy,meta_organ_t_final) %>% 
          mutate(tsne1 = x, tsne2 = y) %>% 
          dplyr::select(-x,-y)
  write.csv(dum,paste0(fl.output,"/tsne_coordination_cca_ccle_",Sys.Date(),".csv"))
```

###plot tsne###
###please edit here for tsne plot###
```{r}
    #set data for tsne
    dum = inner_join(dat_tsne_xy,meta_organ_t_final) %>% 
            filter(organ %in% 
                     c("liver-like"
                       ,"pancreas-like"
                       ,"liver"
                       ,"pancreas"
                     ))
    
    #plot tsne for
    dum = dum %>% 
            dplyr::filter(database %in% "CCLE") %>%
            bind_rows(.
                      ,dum %>% 
                        dplyr::filter(database %in% collection_name) %>% 
                        mutate(abbv_name = unlist(strsplit(full_name
                                                           ,split = "_none")))  %>% 
                        mutate(abbv_name = gsub("[-]","-",abbv_name))
                
                )
     p = dum %>%
        ggplot(.,aes(x = x
                    , y = y
                    , color = organ
                    , label = abbv_name)) +
        geom_point(aes(shape = database
                       , size = 4
                       , fill = organ
                       , alpha = 0.5
        )) +
        geom_text_repel(data = . %>% filter(organ %in% 
                                              c("liver-like"
                                                ,"pancreas-like"))
                        ,aes(x = x
                             , y = y
                             , color = organ
                             , label = abbv_name)
        ) + 
        scale_shape_manual(breaks = c(collection_name,"CCLE")
                           ,values = c(21,24))+
        scale_color_manual(values = c("#50c878"
                                      ,"#50c878"
                                      ,"#842DCE"
                                      ,"#842DCE"
        ))  +
        scale_fill_manual(breaks = c("liver-like"
                                     ,"liver"
                                     ,"pancreas-like"
                                     ,"pancreas")
                          ,values = c("white"
                                      ,"#388c54"
                                      ,"white"
                                      ,"#5c1f90"
                          )) + 
        ggtitle("t-SNE of normalized RNA-seq data") + theme_test() +
        xlab("t-SNE 1") +
        ylab("t-SNE 2")
      
      png(file = paste0(fl.output,"/tsne_ccle_",Sys.Date(),".png")
          ,width = 500
          ,height = 500)
      p
      dev.off()
      
      pdf(file = paste0(fl.output,"/tsne_ccle_",Sys.Date(),".pdf"))
      p
      dev.off()
      
      pdf(file = paste0(fl.output,"/tsne_ccle_",Sys.Date(),".pdf"))
      p
      dev.off()
```

